### ISSUE TITLE
 False positives in `VariableInitialization` around `SetLength` intrinsic

### ISSUE THREAD

**Issue by zedxxx** [integrated-application-development/sonar-delphi#250]
Delphi IDE version: Delphi 12.1

DelphiLint version: 1.1.0

SonarDelphi version: 1.6.0

![1](https://github.com/integrated-application-development/sonar-delphi/assets/6298130/420d064f-14f0-449b-b015-e64e3ca5a132)

```delphi
unit Unit1;

interface

uses
  GR32,
  GR32_Polygons,
  GR32_VectorUtils;

procedure DrawThickLine(ABitmap: TBitmap32; const AStart, AEnd: TFloatPoint; const AColor: TColor32;
  const AWidth: Integer); inline;

procedure DrawThickPolyLine(ABitmap: TBitmap32; const APoints: TArrayOfFloatPoint; const AColor: TColor32;
  const AWidth: Integer);

implementation

procedure DrawThickLine(ABitmap: TBitmap32; const AStart, AEnd: TFloatPoint; const AColor: TColor32;
  const AWidth: Integer);
var
  VLine: TArrayOfFloatPoint;
begin
  SetLength(VLine, 2);

  VLine[0] := AStart;
  VLine[1] := AEnd;

  DrawThickPolyLine(ABitmap, VLine, AColor, AWidth);
end;

procedure DrawThickPolyLine(ABitmap: TBitmap32; const APoints: TArrayOfFloatPoint;
  const AColor: TColor32; const AWidth: Integer);
var
  VPoly: TArrayOfFloatPoint;
begin
  VPoly := BuildPolyLine(APoints, AWidth);

  if not ABitmap.MeasuringMode then begin
    PolygonFS(ABitmap, VPoly, AColor);
  end;

  ABitmap.Changed(MakeRect(PolygonBounds(VPoly), rrOutside));
end;

end.
```

**Comment by Cirras** [comment]
Hi @zedxxx, thanks for reporting these!

The first issue (uninitialized `VLine`) might be a mistake in the way we're modeling the `SetLength` intrinsic.
Definitely a problem - I'll have a look at fixing it.

The second issue (unused `VPoly`) looks like a name resolution issue on the `PolygonFS` invocation - I'm interested in chasing it, but can you make a new GitHub issue to track that distinct bug separately ?

**Comment by zedxxx** [comment]
> The second issue (unused VPoly) looks like a name resolution issue on the PolygonFS invocation - I'm interested in chasing it, but can you make a new GitHub issue to track that distinct bug separately ?

Sure: https://github.com/integrated-application-development/sonar-delphi/issues/251

**Comment by Cirras** [comment]
Cheers, I'll chase up these issues.
Thanks for reporting them. 👍

**Comment by Cirras** [comment]
Hi @zedxxx,

Looks like `SetLength` is modeled correctly - and I can't seem to reproduce this `VariableInitialization` FP.
(See image below)

Here's what I did:
* created a new VCL project
* cloned the `Graphics32` library and added the `Source` folder to the search path in project options
* added your `Unit1` file to the project
* built the project successfully (just to make sure it could compile)
* scanned with DelphiLint (FP issue not reproduced)
  ![image](https://github.com/integrated-application-development/sonar-delphi/assets/22178573/37ab4e34-443d-497b-8a31-b666c9c8d09e)

The analyzer **requires source code** to perform type & invocation resolution, and things can go quite haywire if dependency source code isn't in the search path (or debugger source path).

One potential thing - do you only have the `Graphics32` DCUs on your search path, and not the source files?

**Comment by zedxxx** [comment]
In my case, the path to the sources was added through the Delphi settings, not the project settings. Also, the IDE environment variable is used to specify the path (if it matters).
![1](https://github.com/integrated-application-development/sonar-delphi/assets/6298130/e9ff266f-b9cc-44eb-b215-09a2029be308)

**Comment by Cirras** [comment]
Ah, there's the problem.

The analyzer doesn't currently read the library path. Adding support hasn't been seriously looked at, but it would certainly improve the accuracy of our import resolution (as evidenced by your case).

In your case, for an accurate analysis you'll have to add your dependencies to the search path or debug source path.

We could look into adding support for the library path, and you're welcome to open a feature request.
